cli/src/commands/file/search.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap_complete::ArgValueCompleter;
+use jj_lib::conflicts::MaterializedTreeValue;
+use jj_lib::conflicts::materialize_tree_value;
+use jj_lib::repo::Repo as _;
+use jj_lib::str_util::StringPattern;
+use pollster::FutureExt as _;
+use tracing::instrument;
+
+use crate::cli_util::CommandHelper;
+use crate::cli_util::RevisionArg;
+use crate::cli_util::print_unmatched_explicit_paths;
+use crate::command_error::CommandError;
+use crate::complete;
+use crate::ui::Ui;
+
+/// Search for content in files
+///
+/// Lists files containing the specified pattern.
+///
+/// This is an early version of the command. It only supports substring matching
+/// for now, it doesn't search files concurrently, and it doesn't indicate where
+/// in the file the match was found.
+#[derive(clap::Args, Clone, Debug)]
+pub(crate) struct FileSearchArgs {
+    /// The revision to search files in
+    #[arg(
+        long, short,
+        default_value = "@",
+        value_name = "REVSET",
+        add = ArgValueCompleter::new(complete::revset_expression_all),
+    )]
+    revision: RevisionArg,
+
+    /// The substring to search for
+    #[arg(long, short, value_name = "PATTERN", required = true)]
+    pattern: String,
+
+    /// Only search files matching these prefixes (instead of all files)
+    #[arg(
+        value_name = "FILESETS",
+        value_hint = clap::ValueHint::AnyPath,
+        add = ArgValueCompleter::new(complete::all_revision_files)
+    )]
+    paths: Vec<String>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_file_search(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &FileSearchArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let commit = workspace_command.resolve_single_rev(ui, &args.revision)?;
+    let tree = commit.tree();
+    let fileset_expression = workspace_command.parse_file_patterns(ui, &args.paths)?;
+    let file_matcher = fileset_expression.to_matcher();
+
+    ui.request_pager();
+    let mut formatter = ui.stdout_formatter();
+    let store = workspace_command.repo().store().clone();
+
+    // TODO: Support other patterns than substring
+    let pattern = StringPattern::substring(&args.pattern);
```

Comment:
Maybe better to use `glob()` because it is the new default.

FWIW, `diff_contains()` revset searches for matching lines.
```rust
fn match_lines<'a>(text: &'a [u8], matcher: &StringMatcher) -> impl Iterator<Item = &'a [u8]> {
    // The pattern is matched line by line so that it can be anchored to line
    // start/end. For example, exact:"" will match blank lines.
    text.split_inclusive(|b| *b == b'\n').filter(|line| {
        let line = line.strip_suffix(b"\n").unwrap_or(line);
        matcher.is_match_bytes(line)
    })
}
```